mirrord-layer/src/pod_api.rs

```diff
@@ -417,3 +415,136 @@ async fn create_job_pod_agent(
+        let pod = deployment
+            .spec
+            .and_then(|spec| spec.template.metadata?.name)
+            .map(|pod_name| PodData {
+                pod_name,
+                container_name: None,
+            })?;
+        pod.container_id(pod_api).await
+    }
+}
+
+#[derive(Debug)]
+pub(crate) enum Target {
+    Deployment(DeploymentData),
+    Pod(PodData),
+}
+
+impl Target {
+    pub async fn container_info(
+        &self,
+        pod_api: &Api<Pod>,
+        deployment_api: &Api<Deployment>,
+    ) -> Option<String> {
+        match self {
+            Target::Pod(pod) => pod.container_id(pod_api).await,
+            Target::Deployment(deployment) => {
+                deployment.container_id(deployment_api, pod_api).await
+            }
+        }
+    }
+
+    pub async fn node_name(&self, pods_api: Api<Pod>) -> Option<String> {
+        match self {
+            Target::Pod(PodData {
+                pod_name,
+                container_name: _,
+            }) => {
+                let pod = pods_api.get(pod_name).await.ok()?;
+                pod.spec?.node_name
+            }
+            Target::Deployment(_) => None,
+        }
+    }
+}
+
+impl FromStr for DeploymentData {
+    type Err = LayerError;
+
+    fn from_str(input: &str) -> Result<Self> {
+        let target_data = input.split('/').collect::<Vec<&str>>();
+        match target_data.first() {
+            Some(&"deployment") => Ok(DeploymentData {
+                deployment: target_data[1].to_string(),
+            }),
+            _ => Err(LayerError::InvalidTarget(format!(
+                "Provided target: {:?} is not a deployment.",
+                input
+            ))),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct PodData {
+    pub pod_name: String,
+    pub container_name: Option<String>,
+}
+
+impl FromStr for PodData {
+    type Err = LayerError;
+
+    fn from_str(input: &str) -> Result<Self> {
+        let target_data = input.split('/').collect::<Vec<&str>>();
+        match target_data.first() {
+            Some(&"pod") if target_data.len() == 2 => Ok(PodData {
+                pod_name: target_data[1].to_string(),
+                container_name: None,
+            }),
+            Some(&"pod") if target_data.len() == 4 && target_data[2] == "container" => {
+                Ok(PodData {
+                    pod_name: target_data[1].to_string(),
+                    container_name: Some(target_data[3].to_string()),
+                })
+            }
+            _ => Err(LayerError::InvalidTarget(format!(
+                "Provided target: {:?} is neither a pod or a deployment.",
+                input
+            ))),
+        }
+    }
+}
+
+impl PodData {
+    pub async fn container_id(&self, pods_api: &Api<Pod>) -> Option<String> {
+        let pod = pods_api.get(&self.pod_name).await.ok()?;
+        let container_statuses = &pod.status?.container_statuses?;
+        // TODO: should the return type be a Result here?
```

Comment:
I'm in favor of returning `Result` for functions that can fail, for example on the `pods_api_.get` line, it feels like you're eating up an error?

Reply:
@eyalb181 thoughts? do we want to propagate the error from here or just return None if anything bad happens and just say container not found?